src/Chapter4.hs

```diff
@@ -489,7 +504,42 @@ Implement the 'Applicative' instance for our 'List' type.
   type.
 -}
 
-
+listFlatten :: List a -> List a -> List a
+listFlatten Empty       list2 = list2
+listFlatten list1       Empty = list1
+listFlatten (Cons x xs) list2 = Cons x (listFlatten xs list2)
+
+
+instance Applicative List where
+  pure :: a -> List a
+  pure x = Cons x Empty
+
+  (<*>) :: List (a -> b) -> List a -> List b
+  Empty       <*> _  = Empty
+  (Cons f fs) <*> xs = listFlatten (fmap f xs) (fs <*> xs)
+{- 
+With this solution i was getting the following error, but when I used listFlatten 
+created for Monad is working well :: any assistance on the subject will be great :)
+
+    (Cons f fs) <*> xs = Cons (fmap f xs) (fs <*> xs)
```

Comment:
The result should be a value of type `List b`. When you do `fmap f x` you get a value of type `List b` exactly. Similarly, the type of `fs <*> xs` is also `List b`. So you have two values of type `List b` and you need to produce a value of type `List b`. This is list append, as you implemented :slightly_smiling_face: 

Unlike your `listFlatten` function, the `Cons` operator has the following type (remember, that constructors in Haskell are also just plain functions?)

```haskell
Cons :: a -> List a -> List a
```

So, `Cons` takes a value, a list and produces a list. But if you have two lists of the same type, `Cons` can be applied. You have `Cons :: a -> List a -> List a`, but you apply as if `Cons` had type `Cons :: List b -> List b -> List b`.

The GHC error message is not really helpful here (and GHC is known for not providing beginner-friendly error messages in some cases, unfortunately). But what is means, is that because in Haskell you also have type inference, the compiler tries to infer the type of `Cons` and match it with definition, but it can't.

Reply:
Thanks a lot, 
my doubt is cleared